index.js

```diff
@@ -458,50 +487,57 @@ XiaomiRoborockVacuum.prototype = {
             return;
         }
 
-        log.debug('DEB setCleaning | ' + that.model + ' | Cleaning set it to ' + state + '.');
-        if(state) {
-            if(!that.cleaning) {
+        that.device.cleaning().then((isCleaning) => {
+            that.cleaning = isCleaning;
+            if (state && !that.cleaning) { // Start cleaning
                 log.info('ACT setCleaning | ' + that.model + ' | Start cleaning, not charging.');
-                that.device.activateCleaning();
-                that.cleaning = true
-                that.lastrobotcleaning = that.cleaning;
-                that.charging = false;
-                that.lastrobotcharging = that.charging;
-                that.batteryService.getCharacteristic(Characteristic.ChargingState).updateValue(Characteristic.ChargingState.NOT_CHARGING);
-
-                // Cleaning => leaves dock
-                if (that.dock) {
-                    that.docked = false;
-                    that.dockService.getCharacteristic(Characteristic.OccupancyDetected).updateValue(that.docked);
-                    log.info('ACT setCleaning - dock | ' + that.model + ' | Docked set to %s.', that.docked);
-                }
-
-                // Cleaning => pausing possible
-                if (that.pause) {
-                    that.pausepossible = true;
-                    that.lastrobotpausecleaning = that.pausepossible;
-                    that.pauseService.getCharacteristic(Characteristic.On).updateValue(that.pausepossible);
-                    log.info('ACT setCleaning - pause | ' + that.model + ' | PausePossible set to %s.', that.pausepossible);
-                }
-            } else {
-               log.debug('DEB setCleaning | ' + that.model + ' | Start cleaning not necessary, cleaning already running.');
-            }
-
-        } else {
-            log.info('ACT setCleaning | ' + that.model + ' | Stop cleaning and go to charge.');
-            that.device.activateCharging(); // Charging works for 1st, not for 2nd
-            that.cleaning = false
-            that.lastrobotcleaning = that.cleaning;
-
-            // Cleaning stoped => pausing not possible
-            if (that.pause) {
-                that.pausepossible = false
-                that.lastrobotpausecleaning = that.pausepossible;
-                that.pauseService.getCharacteristic(Characteristic.On).updateValue(that.pausepossible);
+                return that.device.activateCleaning()
+                    // IS THIS REALLY NEEDED?
```

Comment:
I think so. So the status in Homekit is set and we don't wait for the "changed" via the API.

Reply:
That's my question: if the robot doesn't actually start, why should we anticipate it's starting?
On top of that, from HAP implementation of the "set", "get" events: https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Characteristic.js, the value returned in the callback for "get" is the value set for the characteristic (no need to call the updateValue). For the "set" event, resolving the callback without errors makes the characteristic to assume the new value.

I can see it also has the "change" event. Maybe we can use that event for the updates we need to perform to the related Characteristics? When cleaning changes to true, we can enable the pause indicator and the charging state. :)

Reply:
> If the robot doesn't actually start, why should we anticipate it's starting?

The robot must not be started via Homekit, he can also started via the MiHome App.

> I can see it also has the "change" event. Maybe we can use that event for the updates we need to perform to the related Characteristics?

We can. Test it, if it works - change it.

---

At that time I programmed the plugin, the "_change events_" of the miio api sometimes had time delays of up to 30 seconds. Therefore, right after I started / stopped cleaning, I immediately set the pause, dock, and battery statuses. Because I did not like this "delay" in Homekit.

Reply:
I see it now! Thank you for sharing that experience! This weekend I’ll test these scenarios thoroughly:
1. Hook the depending characteristics via the HAP change event so there is 1 place relating these values and see timings
2. Start/stop the cleaning from the Xiaomi app and see how long it takes to update it in HomeKit 
3. Start/stop from the Home app and check how it updates all the other characteristics 
4. Restart Homebridge while the robot is cleaning (and when stopped) and see the values shown for each characteristic

Thank you again!